### PR TITLE
Fix discrete obs support

### LINKED ISSUE #295: [Bug] Multi env problem with Taxi-v3
### 🐛 Bug

The `A2C` and `PPO` algorithms are seeming not working in parallel for the `Taxi-v3` environment. Judging by the error message it is a `numpy` problem. There is no error unless multiprocessing is used. Also, multiprocessing works on other environments such as 'CartPole-v1', so I'm guessing it has to do with the discrete state space.

#### Minimal Code

```python3
from stable_baselines3 import A2C # this can also be PPO, same error
from stable_baselines3.common.env_util import make_vec_env
env = make_vec_env("Taxi-v3", n_envs=4, seed=0)
model = A2C(policy="MlpPolicy", env=env, verbose=True)
model.learn(total_timesteps=10000)
```

#### Traceback

```bash
Using cuda device
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-3-4551c9bb2165> in <module>
      3 env = make_vec_env("Taxi-v3", n_envs=4, seed=0)
      4 model = A2C(policy="MlpPolicy", env=env, verbose=True)
----> 5 model.learn(total_timesteps=10_000)

~/envs/bsh/lib/python3.8/site-packages/stable_baselines3/a2c/a2c.py in learn(self, total_timesteps, callback, log_interval, eval_env, eval_freq, n_eval_episodes, tb_log_name, eval_log_path, reset_num_timesteps)
    181     ) -> "A2C":
    182 
--> 183         return super(A2C, self).learn(
    184             total_timesteps=total_timesteps,
    185             callback=callback,

~/envs/bsh/lib/python3.8/site-packages/stable_baselines3/common/on_policy_algorithm.py in learn(self, total_timesteps, callback, log_interval, eval_env, eval_freq, n_eval_episodes, tb_log_name, eval_log_path, reset_num_timesteps)
    220         while self.num_timesteps < total_timesteps:
    221 
--> 222             continue_training = self.collect_rollouts(self.env, callback, self.rollout_buffer, n_rollout_steps=self.n_steps)
    223 
    224             if continue_training is False:

~/envs/bsh/lib/python3.8/site-packages/stable_baselines3/common/on_policy_algorithm.py in collect_rollouts(self, env, callback, rollout_buffer, n_rollout_steps)
    176                 # Reshape in case of discrete action
    177                 actions = actions.reshape(-1, 1)
--> 178             rollout_buffer.add(self._last_obs, actions, rewards, self._last_dones, values, log_probs)
    179             self._last_obs = new_obs
    180             self._last_dones = dones

~/envs/bsh/lib/python3.8/site-packages/stable_baselines3/common/buffers.py in add(self, obs, action, reward, done, value, log_prob)
    350             log_prob = log_prob.reshape(-1, 1)
    351 
--> 352         self.observations[self.pos] = np.array(obs).copy()
    353         self.actions[self.pos] = np.array(action).copy()
    354         self.rewards[self.pos] = np.array(reward).copy()

ValueError: could not broadcast input array from shape (4) into shape (4,1)
```

### Expected behavior

A model to get trained in parallel using A2C (or PPO).

### System Info

* pip inside venv
* Graphics card info: GeForce GTX 1050 Ti, Nvidia driver version 460.27.04, CUDA version 11.2
* Python version 3.8.7
* PyTorch version 1.7.1
* Gym version 0.18.0

### Additional context

No additional context.

### Checklist

- [x] I have checked that there is no similar [issue](https://github.com/DLR-RM/stable-baselines3/issues) in the repo (**required**)
- [x] I have read the [documentation](https://stable-baselines3.readthedocs.io/en/master/) (**required**)
- [x] I have provided a minimal working example to reproduce the bug (**required**)

### PR DESCRIPTION
**Pull request by araffin**
## Description

Fixed discrete obs support when using multple env with A2C/PPO.

I also fixed some warnings in the tests (truncated minibatches).

## Motivation and Context

- [x] I have raised an issue to propose this change ([required](https://github.com/DLR-RM/stable-baselines3/blob/master/CONTRIBUTING.md) for new features and bug fixes)

closes #295 

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)

## Checklist:

- [x] I've read the [CONTRIBUTION](https://github.com/DLR-RM/stable-baselines3/blob/master/CONTRIBUTING.md) guide (**required**)
- [x] I have updated the changelog accordingly (**required**).
- [ ] My change requires a change to the documentation.
- [x] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [ ] I have updated the documentation accordingly.
- [x] I have reformatted the code using `make format` (**required**)
- [x] I have checked the codestyle using `make check-codestyle` and `make lint` (**required**)
- [x] I have ensured `make pytest` and `make type` both pass. (**required**)
- [x] I have checked that the documentation builds using `make doc` (**required**)

Note: You can run most of the checks using `make commit-checks`.

Note: we are using a maximum length of 127 characters per line

### CODE REVIEW

File: tests/test_cnn.py [hunk 1]
```diff
@@ -21,7 +21,7 @@ def test_cnn(tmp_path, model_class):
     # to check that the network handle it automatically
     env = FakeImageEnv(screen_height=40, screen_width=40, n_channels=1, discrete=model_class not in {SAC, TD3})
     if model_class in {A2C, PPO}:
-        kwargs = dict(n_steps=100)
+        kwargs = dict(n_steps=100, batch_size=100)
```
Comment (Miffyli) [inline tests/test_cnn.py:24]: This seems to break the test. I would fix this but I am not sure what the intention here was.

**Comment by Miffyli** [comment]
I pushed a suggested edit. The buffer situation was a symptom of deeper cause inside vec envs and how they treat spaces (assume `space.shape` is always good enough).

**Comment by Miffyli** [comment]
Okey never mind, should have known that would break everything horribly. I still find the change only to buffers rather suspicious. Instead, a fix where vecenv returns (n_envs, 1) observations for discrete environments would be nicer.

Using `get_obs_shape` function did not work (broke a ton of tests), but what happens if you just add a special case for discrete spaces?

File: tests/test_cnn.py (hunk 1 above)
Comment (araffin) [reply tests/test_cnn.py:24]: I forgot A2C did not have a "batch_size" argument... this was to remove a warning about truncated mini-batch, I will change to `n_steps=64` instead

**Comment by araffin** [comment]
>Okey never mind, should have known that would break everything horribly. I still find the change only to buffers rather suspicious. Instead, a fix where vecenv returns (n_envs, 1) observations for discrete environments would be nicer.

I would agree on that.
So, the failure is due to a very unfortunate case where numpy does not want to broadcast `(2,)` to `(2, 1)` (where it works for `(1,)` to `(1, 1)`... )

>Using get_obs_shape function did not work (broke a ton of tests), but what happens if you just add a special case for discrete spaces?

Where do you want to add that special case exactly?

**Comment by araffin** [comment]
I've moved the reshape to a special case for discrete obs ;)

**Review by Miffyli — APPROVED** [review]
LGTM with this change.

**Pull request merged**